File: Newton's Method for Roots/newtoncube.py
# Created by Brian Mascitello to calculate x^3 polynomials and below.

def newtoncube(a, b, c, d, x0):
    """ function newton(a, b, c, d, x0) finds a root of the nonlinear function
        specified by f and fprime. y = a*x^3 + b*x^2 + c*x + d;
        yprime = 3*a*x^2 + 2*b*x + c; Result x is the root. """

    epsilon = 2.2204*10**-16
    """ governs precision of convergence
        where 2.2204*10**-16 = machine epsilion in python """
        
    print "f(x) = %dx^3 + %dx^2 + %dx + %d" % (a, b, c, d)

    x = x0
    xprevious = 0
    k = 0

    while abs(float(a*x**3 + b*x**2 + c*x + d)) > epsilon*abs(float(a*x0**3 + b*x0**2 + c*x0 + d)) and k < 20:
        k = k+1
        xprevious = x
        x = float(x) - (float(a*x**3 + b*x**2 + c*x + d)/float(3*a*x**2 + 2*b*x + c))
        change = abs(float(x - xprevious))
        residual = a*x**3 + b*x**2 + c*x + d
        print "Iteration: %d, Root: %f, Change: %f, Residual: %f" % (k,x,change,residual)
    
    print "Root at",x,"\n"

    return float(x)

newtoncube(3, -2, 7, -1, 10)
newtoncube(1, -5, 0, -3, 10)

File: Newton's Method for Roots/newtoncube.m
% Modified by Brian Mascitello to calculate x^3 polynomials
% and below.

function newtoncube(a, b, c, d, x0)
% function newton(a, b, c, d, x0) finds a root of the
% nonlinear function specified by f and fprime.
% y = a*x^3 + b*x^2 + c*x + d;
% yprime = 3*a*x^2 + 2*b*x + c;
% Result x is the root.
% Set Command Window to "format long" to see longer x values.

epsilon = eps; % governs precision of convergence where
% eps = machine epsilion in MATLAB

% sometimes can't achieve eps; can take epsilon = 10^-6 for example
x = x0;
xprev = 0;
k = 0;

% while abs(x-xprev) > epsilon*abs(x) && k < 20
while abs(f(x)) > epsilon*abs(f(x0)) && k < 20
    k = k+1
    xprev = x;
    x = x - f(x)/fprime(x)
    change = abs(x-xprev)
    residual = f(x)
end

    function y = f(x)
    y = a*x^3 + b*x^2 + c*x + d;
    end

    function yprime = fprime(x)
    yprime = 3*a*x^2 + 2*b*x + c;
    end

end